clients/src/test/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandlerTest.java

```diff
@@ -0,0 +1,308 @@
+        Errors error = Errors.UNKNOWN_SERVER_ERROR;
+        Map<TopicPartition, Short> errorsByPartition = new HashMap<>();
+        errorsByPartition.put(errorPartition, error.code());
+
+        ApiResult<TopicPartition, ListOffsetsResultInfo> result =
+            handleResponse(createResponse(errorsByPartition));
+
+        Map<TopicPartition, Throwable> failed = new HashMap<>();
+        failed.put(errorPartition, error.exception());
+        Set<TopicPartition> completed = new HashSet<>(offsetTimestampsByPartition.keySet());
+        completed.removeAll(failed.keySet());
+        assertResult(result, completed, failed, emptyList(), emptySet());
+    }
+
+    @Test
+    public void testHandleResponseSanityCheck() {
+        TopicPartition errorPartition = t0p0;
+        Map<TopicPartition, Long> specsByPartition = new HashMap<>(offsetTimestampsByPartition);
+        specsByPartition.remove(errorPartition);
+
+        ApiResult<TopicPartition, ListOffsetsResultInfo> result =
+            handleResponse(createResponse(emptyMap(), specsByPartition));
+
+        assertEquals(offsetTimestampsByPartition.size() - 1, result.completedKeys.size());
+        assertEquals(1, result.failedKeys.size());
+        assertEquals(errorPartition, result.failedKeys.keySet().iterator().next());
+        String sanityCheckMessage = result.failedKeys.get(errorPartition).getMessage();
+        assertTrue(sanityCheckMessage.contains("did not contain a result for topic partition"));
+        assertTrue(result.unmappedKeys.isEmpty());
+    }
+
+    @Test
+    public void testHandleResponseUnsupportedVersion() {
+        UnsupportedVersionException uve = new UnsupportedVersionException("");
+        Map<TopicPartition, OffsetSpec> maxTimestampPartitions = new HashMap<>();
+        maxTimestampPartitions.put(t1p1, OffsetSpec.maxTimestamp());
+
+        ListOffsetsHandler handler =
+            new ListOffsetsHandler(offsetTimestampsByPartition, new ListOffsetsOptions(), logContext);
+
+        final Map<TopicPartition, Long> nonMaxTimestampPartitions = new HashMap<>(offsetTimestampsByPartition);
+        maxTimestampPartitions.forEach((k, v) -> nonMaxTimestampPartitions.remove(k));
+        // Unsupported version exceptions currently cannot be handled if there's no partition with a
+        // MAX_TIMESTAMP spec...
+        Set<TopicPartition> keysToTest = nonMaxTimestampPartitions.keySet();
+        Set<TopicPartition> expectedFailures = keysToTest;
+        assertEquals(
+            mapToError(expectedFailures, uve),
+            handler.handleUnsupportedVersionException(uve, keysToTest));
+
+        // ...or if there are only partitions with MAX_TIMESTAMP specs...
+        keysToTest = maxTimestampPartitions.keySet();
+        expectedFailures = keysToTest;
+        assertEquals(
+            mapToError(expectedFailures, uve),
+            handler.handleUnsupportedVersionException(uve, keysToTest));
+
+        keysToTest = offsetTimestampsByPartition.keySet();
```

Comment:
nit: add a comment about this part, for example, xxx partitions both MAX_TIMESTAMP and not.